games/monop/monop.6
.\"	$NetBSD: monop.6,v 1.18.2.1 2008/03/24 07:14:42 keiichi Exp $
.\"
.\" Copyright (c) 1980 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)monop.6	6.5 (Berkeley) 3/25/93
.\"
.Dd February 23, 2008
.Dt MONOP 6
.Os
.Sh NAME
.Nm monop
.Nd Monopoly game
.Sh SYNOPSIS
.Nm
.Op Ar file
.Sh DESCRIPTION
.Nm
is reminiscent of the Parker Brother's game Monopoly, and
monitors a game between 2 to 9 users.
It is assumed that the rules of Monopoly are known.
The game follows the standard rules, with the exception that,
if a property goes up for auction and there are only two solvent players,
no auction is held and the property remains unowned.
.Pp
The game, in effect, lends the player money,
so it is possible to buy something which you cannot afford.
However, as soon as a person goes into debt, he must
.Dq fix the problem ,
i.e., make himself solvent, before play can continue.
If this is not possible, the player's property reverts to his debtee,
either a player or the bank.
A player can resign at any time to any person or the bank,
which puts the property back on the board, unowned.
.Pp
Any time that the response to a question is a
.Em string ,
e.g., a name, place or person, you can type
.Sq Ic \&?
to get a list of valid answers.
It is not possible to input a negative number, nor is it ever necessary.
.Pp
.Em A Summary of Commands :
.Bl -tag -width item
.It Ic quit
quit game: This allows you to quit the game.
It asks you if you're sure.
.It Ic print
print board: This prints out the current board.
The columns have the following meanings (column headings are the same for the
.Ic where ,
.Ic own holdings ,
and
.Ic holdings
commands):
.Pp
.Bl -tag -width indent -compact
.It Name
The first ten characters of the name of the square.
.It Own
The
.Em number
of the owner of the property.
.It Price
The cost of the property (if any).
.It Mg
This field has a
.Sq *
in it if the property is mortgaged.
.It \&#
If the property is a Utility or Railroad, this is the number
of such owned by the owner.
If the property is land, this is the number of houses on it.
.It Rent
Current rent on the property.
If it is not owned, there is no rent.
.El
.It Ic where
where players are: Tells you where all the players are.
A
.Sq *
indicates the current player.
.It Ic own holdings
List your own holdings,
i.e., money, get-out-of-jail-free cards, and property.
.It Ic holdings
holdings list: Look at anyone's holdings.
It will ask you whose holdings you wish to look at.
When you are finished, type
.Sq Ic done .
.It Ic mortgage
mortgage property:
Sets up a list of mortgageable property, and asks which you wish to mortgage.
.It Ic unmortgage
unmortgage property:
Unmortgage mortgaged property.
.It Ic buy
buy houses:
Sets up a list of monopolies on which you can buy houses.
If there is more than one, it asks you which you want to buy for.
It then asks you how many for each piece of property,
giving the current amount in parentheses after the property name.
If you build in an unbalanced manner
(a disparity of more than one house within the same monopoly),
it asks you to re-input things.
.It Ic sell
sell houses:
Sets up a list of monopolies from which you can sell houses.
It operates in an analogous manner to
.Ic buy .
.It Ic card
card for jail:
Use a get-out-of-jail-free card to get out of jail.
If you're not in jail, or you don't have one, it tells you so.
.It Ic pay
pay for jail:
Pay $50 to get out of jail, from whence you are put on Just Visiting.
Difficult to do if you're not there.
.It Ic trade
This allows you to trade with another player.
It asks you whom you wish to trade with,
and then asks you what each wishes to give up.
You can get a summary at the end, and, in all cases,
it asks for confirmation of the trade before doing it.
.It Ic resign
Resign to another player or the bank.
If you resign to the bank, all property reverts to its virgin state,
and get-out-of-jail-free cards revert to the deck.
.It Ic save
save game:
Save the current game in a file for later play.
You can continue play after saving,
either by adding the file in which you saved the game after the
.Nm
command, or by using the
.Ic restore
command (see below).
It will ask you which file you wish to save it in,
and, if the file exists, confirm that you wish to overwrite it.
.It Ic restore
restore game:
Read in a previously saved game from a file.
It leaves the file intact.
.It Ic roll
Roll the dice and move forward to your new location.
If you simply hit the
.Aq RETURN
key instead of a command, it is the same as typing
.Ic roll .
.El
.Sh AUTHORS
.An Ken Arnold
.Sh BUGS
No command can be given an argument instead of a response to a query.


doc/CHANGES
LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.1035 $>


[Note: This file does not mention every change made to the NetBSD source tree.
For an exhaustive list, please subscribe to the `source-changes' mailing list,
or see the mailing list archives.  For more information, send a message
containing just the word `help' to `majordomo@NetBSD.org'.]

------------------------------------------------------------------------------
PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:

	theme: Content ...
		more content ...
		more content. [developer DATE-FORMAT]

That's it, one tab to start an item, and two tabs to indent the content.
Also please start the content with a capital and end it with a dot.

This is very important for the www scripts, otherwise it won't be able to
parse the file and the HTML output will be mangled.
------------------------------------------------------------------------------


Changes from NetBSD 4.0 to NetBSD 5.0:
	evbppc: Add support for Xilinx Virtex II-Pro/4-FX. [freza 20061202]
	mac68k: Add support for four more Nubus video cards, including the
		VillageTronic MacPicasso 340. [hauke 20061206]
	ixp425: Add support for NPE Ethernet. [scw 20061210]
	nfs: Remove NQNFS support from kernel. [yamt 20061227]
	mount_psshfs(8): Add puffs sshfs. [pooka 20061229]
	zaurus: Add support for Sharp C3x00 PDA. [ober 20070105]
	i386: Add support for Microsoft Xbox. [jmcneill 20070105]
	ntp: Import ntp 4.2.4 (20061228). Tracking of network interface
		address changes is now supported - no more ntpd
		restarts when interface addresses are re-configured
		(WLAN/DSL/ppp/...). [kardel 20070106]
	kernel: Global 'securelevel' variable removed; it is now internal to
		the secmodel. [elad 20070110]
	gcc(1): Updated to version 4.1.2-20070110. [mrg 20070111]
	acpitools: Added from FreeBSD via fvdl. [christos 20070114]
	scsictl(8): Added setspeed command, to set the highest speed that the
		optical drive should use for reading data,
		contributed by Kimura Fuyuki. [bouyer 20070122]
	i386, amd64: Add cmos(4), driver for x86 CMOS RAM. [dyoung 20070206]
	newlock2: Merge the newlock2 branch. Introduces a number of new
		kernel synchronization primitives, improves scalability
		on MP systems, and replaces the existing SA threads model
		with a 1:1 threading model. [ad 20070209]
	news68k: Implement generic softintr(9) support. [tsutsui 20070210]
	refuse: Added FUSE (file system in userspace) compatibility
		functionality, layered on top of libpuffs. [agc 20070211]
	i386: Added an opencrypto provider for VIA ACE (AES encryption
		instructions), ported from OpenBSD. [daniel 20070217]
	sdiff(1): Replaced by OpenBSD sdiff(1) (public domain). [rmind 20070219]
	btuart(4): Add support for Bluetooth HCI UART (H4) driver.
		[kiyohara 20070221]
	btuartd(8): Add support for Bluetooth HCI UART (H4) daemon.
		[kiyohara 20070221]
	mutex(9), rwlock(9): Support priority inheritance. [yamt 20070226]
	eeprom(8): Add the missing kernel bits on arm and powerpc that kept
		eeprom from actually changing firmware settings.
		[macallan 20070228]
	MAKEDEV(8), MAKEDEV.local(8): Let MAKEDEV.local use functions
		defined in MAKEDEV. [apb 20070303]
	x68k: Add generic softintr(9) support. [tsutsui 20070304]
	luna68k: Add generic softintr(9) support. [tsutsui 20070304]
	file(1): Update to 4.20.  [pooka 20070304]
	hfs: Add read-only Apple HFS+ file system support. [dillo 20070305]
	ipv6: Add ipv6 fast forward. [liamjfoy 20070307]
	openssh: Import 4.6. [christos 20070310]
	aiboost(4): ASUS AI Booster ACPI Hardware monitor driver,
		adapted from FreeBSD for envsys(4). [xtraeme 20070313]
	evbmips: Add Infineon ADM5120 port. [dyoung 20070320]
	i386, amd64: Intel(R) On Demand Clock Modulation driver, to save
		power and overheating in CPUs supporting the Thermal
		Monitor feature (TM). Adapted from OpenBSD/FreeBSD.
		See options(4) for more information. [xtraeme 20070320]
	kernel: Import of DRM (Direct Rendering Manager). Not yet enabled
		by default. [drochner 20070320]
	libc: Add support for LC_TIME. [manu 20070328]
	genfb(4): Generic framebuffer console driver with PCI and SBus
		frontends. [macallan 20070411]
	sgimips: Add support for SGI O2 PS/2 controller macekbc(4) and
		onboard display adapter crmfb(4). [jmcneill 20070412]
	sgimips: Import mavb(4) audio device driver for SGI O2 from
		OpenBSD. [jmcneill 20070412]
	sparc64: Add support for hardware monitoring on
		Ultra Enterprise 450. See envctrl(4). [tnn 20070414]
	puffs(4): Support NFS exporting puffs file servers. [pooka 20070416]
	xargs(1): Bring in xargs from FreeBSD to gain -I, -J etc. but keep our
		GNU compatible exit values. [christos 20070418]
	mount_9p(8): Add support to mount a 9P file service with puffs.
		[pooka 20070421]
	kernel: Remove the CHECK_ALT_xxx() defines and emul_find() used
		to locate files within an emulation's root directory.
		Instead get namei() to check the emulation root first.
		[dsl 20070422]
	prep: Add support for the MTX604 RAVEN based machine. Other RAVEN
		based machines are also likely to work now. [garbled 20070426]
	kernel: Added support for POSIX Asynchronous I/O. [rmind 20070430]
	puffs(3): Add puffs_framebuf(3), which abstracts the buffer
		management and event loop for networked file systems
		using puffs_cc(3). Convert mount_psshfs(8) and
		mount_9p(8) to use it. [pooka 20070505]
	kernel: Add the TFTPROOT kernel option for TFTP'ing root RAMdisk
		at root mount time. [manu 20070508]
	pf(4): Make 'nat' and 'rdr' translation rules obey state
		policy flags ('if-bound', 'group-bound', 'floating').
		Extend pf.conf(5) syntax to express the translation-state
		policies. [dyoung 20070510]
	idlelwp: yamt-idlelwp branch was merged.
		1. separate context switching and thread scheduling.
		2. implement idle lwp.
		3. clean up related MD/MI interfaces.
		4. make scheduler(s) modular.
		[yamt 20070518]
	ld.elf_so(1): Support Solaris $ORIGIN etc. expansions in paths.
		[christos 20070518]

games/monop/monop.c
/*	$NetBSD: monop.c,v 1.18.2.1 2008/03/24 07:14:42 keiichi Exp $	*/

/*
 * Copyright (c) 1980, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1980, 1993\n\
	The Regents of the University of California.  All rights reserved.\n");
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)monop.c	8.1 (Berkeley) 5/31/93";
#else
__RCSID("$NetBSD: monop.c,v 1.18.2.1 2008/03/24 07:14:42 keiichi Exp $");
#endif
#endif /* not lint */

#include <stdio.h>
#include <signal.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>
#include "deck.h"
#include "monop.h"

int main(int, char *[]);
static void getplayers(void);
static void init_players(void);
static void init_monops(void);
static void do_quit(int);


bool	fixing,			/* set if fixing up debt		*/
	trading,		/* set if in process of trading		*/
	told_em,		/* set if told user he's out of debt	*/
	spec;			/* set if moving by card to RR or UTIL	*/

const char	*name_list[MAX_PL+2],	/* list of players' names	*/
	*const comlist[]	= {	/* list of normal commands 	*/
	"quit",		/*  0 */	"print",	/*  1 */
	"where",	/*  2 */	"own holdings",	/*  3 */
	"holdings",	/*  4 */	"mortgage",	/*  5 */
	"unmortgage",	/*  6 */	"buy houses",	/*  7 */
	"sell houses",	/*  8 */	"card",		/*  9 */
	"pay",		/* 10 */	"trade",	/* 11 */
	"resign",	/* 12 */	"save",		/* 13 */
	"restore",	/* 14 */	"roll",		/* 15 */
	"",		/* 16 */
	0
	},
	*const yncoms[]	= {	/* list of commands for yes/no answers	*/
	"yes",		/*  0 */	"no",		/*  1 */
	"quit",		/*  2 */	"print",	/*  3 */
	"where",	/*  4 */	"own holdings",	/*  5 */
	"holdings",	/*  6 */
	0
	},
	*const lucky_mes[]	= {	/* "got lucky" messages		*/
	"You lucky stiff",		"You got lucky",
	"What a lucky person!",		"You must have a 4-leaf clover",
	"My, my!  Aren't we lucky!",	"Luck smiles upon you",
	"You got lucky this time",	"Lucky person!",
	"Your karma must certainly be together",
	"How beautifully Cosmic",	"Wow, you must be really with it"
	/* "I want your autograph",	-- Save for later */
	};

int	player,			/* current player number		*/
	num_play,		/* current number of players		*/
	num_doub,		/* # of doubles current player rolled	*/
				/* # of "got lucky" messages		*/
	num_luck	= sizeof lucky_mes / sizeof (char *);

/* list of command functions		*/
void (*const func[])(void) = { /* array of function calls for commands */
	quit,			/* quit game		|*  0 *|	*/
	printboard,		/* print board		|*  1 *|	*/
	where,			/* where players are	|*  2 *|	*/
	list,			/* own holdings		|*  3 *|	*/
	list_all,		/* holdings list	|*  4 *|	*/
	mortgage,		/* mortgage property	|*  5 *|	*/
	unmortgage,		/* unmortgage property	|*  6 *|	*/
	buy_houses,		/* buy houses		|*  7 *|	*/
	sell_houses,		/* sell houses		|*  8 *|	*/
	card,			/* card for jail	|*  9 *|	*/
	pay,			/* pay for jail		|* 10 *|	*/
	trade,			/* trade		|* 11 *|	*/
	resign,			/* resign		|* 12 *|	*/
	save,			/* save game		|* 13 *|	*/
	restore,		/* restore game		|* 14 *|	*/
	do_move,		/* roll			|* 15 *|	*/
	do_move			/* ""			|* 16 *|	*/
	};

DECK	deck[2];		/* Chance and Community Chest		*/

PLAY	*play,			/* player structure array ("calloc"ed)	*/
	*cur_p;			/* pointer to current player's struct	*/

RR_S	rr[N_RR];		/* railroad descriptions		*/

UTIL_S	util[2];		/* utility descriptions			*/

#define MONINIT(num_in, h_cost, not_m, mon_n, sq1,sq2,sq3) \
     {0,    -1, num_in, 0,      h_cost, not_m, mon_n, {sq1,sq2,sq3}, {0,0,0}}
/* name  owner          num_own                                      sq */

MON	mon[N_MON]	= {	/* monopoly descriptions		*/
/*   num_in h_cost  not_m	mon_n	    sqnums */
MONINIT(2,  1,	"Purple",	"PURPLE",   1,3, 0),
MONINIT(3,  1,	"Lt. Blue",	"LT. BLUE", 6,8,9),
MONINIT(3,  2,	"Violet",	"VIOLET",   11,13,14),
MONINIT(3,  2,	"Orange",	"ORANGE",   16,18,19),
MONINIT(3,  3,	"Red",		"RED",	    21,23,24),
MONINIT(3,  3,	"Yellow",	"YELLOW",   26,27,29),
MONINIT(3,  4,	"Green",	"GREEN",    31,32,34),
MONINIT(2,  4,	"Dk. Blue",	"DK. BLUE", 37,39, 0),
};
#undef MONINIT

PROP	prop[N_PROP]	= {	/* typical properties			*/
/* morg	monop	square	houses	mon_desc	rent	*/
{0,	0,	1,	0,	&mon[0],	{ 2, 10, 30,  90, 160, 250} },
{0,	0,	3,	0,	&mon[0],	{ 4, 20, 60, 180, 320, 450} },
{0,	0,	6,	0,	&mon[1],	{ 6, 30, 90, 270, 400, 550} },
{0,	0,	7,	0,	&mon[1],	{ 6, 30, 90, 270, 400, 550} },
{0,	0,	9,	0,	&mon[1],	{ 8, 40,100, 300, 450, 600} },
{0,	0,	11,	0,	&mon[2],	{10, 50,150, 450, 625, 750} },
{0,	0,	13,	0,	&mon[2],	{10, 50,150, 450, 625, 750} },
{0,	0,	14,	0,	&mon[2],	{12, 60,180, 500, 700, 900} },
{0,	0,	16,	0,	&mon[3],	{14, 70,200, 550, 750, 950} },
{0,	0,	17,	0,	&mon[3],	{14, 70,200, 550, 750, 950} },
{0,	0,	19,	0,	&mon[3],	{16, 80,220, 600, 800,1000} },
{0,	0,	21,	0,	&mon[4],	{18, 90,250, 700, 875,1050} },
{0,	0,	23,	0,	&mon[4],	{18, 90,250, 700, 875,1050} },
{0,	0,	24,	0,	&mon[4],	{20,100,300, 750, 925,1100} },
{0,	0,	26,	0,	&mon[5],	{22,110,330, 800, 975,1150} },
{0,	0,	27,	0,	&mon[5],	{22,110,330, 800, 975,1150} },
{0,	0,	29,	0,	&mon[5],	{24,120,360, 850,1025,1200} },
{0,	0,	31,	0,	&mon[6],	{26,130,390, 900,1100,1275} },
{0,	0,	32,	0,	&mon[6],	{26,130,390, 900,1100,1275} },
{0,	0,	34,	0,	&mon[6],	{28,150,450,1000,1200,1400} },
{0,	0,	37,	0,	&mon[7],	{35,175,500,1100,1300,1500} },
{0,	0,	39,	0,	&mon[7],	{50,200,600,1400,1700,2000} }
};

SQUARE	board[N_SQRS+1]	= {	/* board itself (+1 for Jail)		*/
/* name (COLOR)			owner	type	desc		cost	*/

{"=== GO ===",			-1,	SAFE,	NULL,		0	},
{"Mediterranean Ave. (P)",	-1,	PRPTY,	&prop[0],	60	},
{"Community Chest i",		-1,	CC,	NULL,		0	},
{"Baltic Ave. (P)",		-1,	PRPTY,	&prop[1],	60	},
{"Income Tax",			-1,	INC_TAX, NULL,		0	},
{"Reading RR",			-1,	RR,	&rr[0],		200	},
{"Oriental Ave. (L)",		-1,	PRPTY,	&prop[2],	100	},
{"Chance i",			-1,	CHANCE,	NULL,		0	},
{"Vermont Ave. (L)",		-1,	PRPTY,	&prop[3],	100	},
{"Connecticut Ave. (L)",	-1,	PRPTY,	&prop[4],	120	},
{"Just Visiting",		-1,	SAFE,	NULL,		0	},
{"St. Charles Pl. (V)",		-1,	PRPTY,	&prop[5],	140	},
{"Electric Co.",		-1,	UTIL,	&util[0],	150	},
{"States Ave. (V)",		-1,	PRPTY,	&prop[6],	140	},
{"Virginia Ave. (V)",		-1,	PRPTY,	&prop[7],	160	},
{"Pennsylvania RR",		-1,	RR,	&rr[1],		200	},
{"St. James Pl. (O)",		-1,	PRPTY,	&prop[8],	180	},
{"Community Chest ii",		-1,	CC,	NULL,		0	},
{"Tennessee Ave. (O)",		-1,	PRPTY,	&prop[9],	180	},
{"New York Ave. (O)",		-1,	PRPTY,	&prop[10],	200	},
{"Free Parking",		-1,	SAFE,	NULL,		0	},
{"Kentucky Ave. (R)",		-1,	PRPTY,	&prop[11],	220	},
{"Chance ii",			-1,	CHANCE,	NULL,		0	},
{"Indiana Ave. (R)",		-1,	PRPTY,	&prop[12],	220	},
{"Illinois Ave. (R)",		-1,	PRPTY,	&prop[13],	240	},
{"B&O RR",			-1,	RR,	&rr[2],		200	},
{"Atlantic Ave. (Y)",		-1,	PRPTY,	&prop[14],	260	},
{"Ventnor Ave. (Y)",		-1,	PRPTY,	&prop[15],	260	},
{"Water Works",			-1,	UTIL,	&util[1],	150	},
{"Marvin Gardens (Y)",		-1,	PRPTY,	&prop[16],	280	},
{"GO TO JAIL",			-1,	GOTO_J,	NULL,		0	},
{"Pacific Ave. (G)",		-1,	PRPTY,	&prop[17],	300	},
{"N. Carolina Ave. (G)",	-1,	PRPTY,	&prop[18],	300	},
{"Community Chest iii",		-1,	CC,	NULL,		0	},
{"Pennsylvania Ave. (G)",	-1,	PRPTY,	&prop[19],	320	},
{"Short Line RR",		-1,	RR,	&rr[3],		200	},
{"Chance iii",			-1,	CHANCE,	NULL,		0	},
{"Park Place (D)",		-1,	PRPTY,	&prop[20],	350	},
{"Luxury Tax",			-1,	LUX_TAX, NULL,		0	},
{"Boardwalk (D)",		-1,	PRPTY,	&prop[21],	400	},
{"JAIL",			-1,	IN_JAIL, NULL,		0	}
};


/*
 *	This program implements a monopoly game
 */
int
main(ac, av)
	int ac;
	char *av[];
{
	/* Revoke setgid privileges */
	setgid(getgid());

	srandom((unsigned long)time(NULL));
	num_luck = sizeof lucky_mes / sizeof (char *);
	init_decks();
	init_monops();
	if (ac > 1) {
		if (rest_f(av[1]) < 0)
			restore();
	}
	else {
		getplayers();
		init_players();
	}
	signal(SIGINT, do_quit);
	for (;;) {
		printf("\n%s (%d) (cash $%d) on %s\n", cur_p->name, player + 1,
			cur_p->money, board[cur_p->loc].name);
		printturn();
		force_morg();
		execute(getinp("-- Command: ", comlist));
	}
}

/*ARGSUSED*/
static void
do_quit(n)
	int n __unused;
{
	quit();
}

/*
 *	This routine gets the names of the players
 */
static void
getplayers()
{
	int i, j;
	char buf[257];

blew_it:
	for (;;) {
		if ((num_play = get_int("How many players? ")) <= 1 ||
		    num_play > MAX_PL)
			printf("Sorry. Number must range from 2 to %d\n",
			    MAX_PL);
		else
			break;
	}
	cur_p = play = calloc((size_t)num_play, sizeof (PLAY));
	if (play == NULL)
		err(1, NULL);
	for (i = 0; i < num_play; i++) {
		do {
			printf("Player %d's name: ", i + 1);
			fgets(buf, sizeof(buf), stdin);
			if (feof(stdin)) {
				printf("End of file on stdin\n");
				exit(0);
			}
			buf[strcspn(buf, "\n")] = '\0';
		} while (strlen(buf) == 0);
		name_list[i] = play[i].name = strdup(buf);
		if (name_list[i] == NULL)
			err(1, NULL);
		play[i].money = 1500;
	}
	name_list[i++] = "done";
	name_list[i] = 0;
	for (i = 0; i < num_play; i++)
		for (j = i + 1; j <= num_play; j++)
			if (strcasecmp(name_list[i], name_list[j]) == 0) {
				if (j != num_play)
					printf("Hey!!! Some of those are "
					    "IDENTICAL!!  Let's try that "
					    "again...\n");
				else
					printf("\"done\" is a reserved word.  "
					    "Please try again\n");
				for (i = 0; i < num_play; i++)
					free(play[i].name);
				free(play);
				goto blew_it;
			}
}

/*
 *	This routine figures out who goes first
 */
static void
init_players()
{
	int i, rl, cur_max;
	bool over = 0;
	int max_pl = 0;

again:
	putchar('\n');
	for (cur_max = i = 0; i < num_play; i++) {
		printf("%s (%d) rolls %d\n", play[i].name, i+1, rl=roll(2, 6));
		if (rl > cur_max) {
			over = FALSE;
			cur_max = rl;
			max_pl = i;
		}
		else if (rl == cur_max)
			over++;
	}
	if (over) {
		printf("%d people rolled the same thing, so we'll try again\n",
		    over + 1);
		goto again;
	}
	player = max_pl;
	cur_p = &play[max_pl];
	printf("%s (%d) goes first\n", cur_p->name, max_pl + 1);
}

/*
 *	This routine initializes the monopoly structures.
 */
static void
init_monops()
{
	MON *mp;
	int i;

	for (mp = mon; mp < &mon[N_MON]; mp++) {
		mp->name = mp->not_m;
		for (i = 0; i < mp->num_in; i++)
			mp->sq[i] = &board[mp->sqnums[i]];
	}
}
